### PR TITLE
Fix #586 (display the killing clipping rectangle)

### LINKED ISSUE #586: Mark corner of map by kill tiles
You are not able to go out of the map, but you also cant know when you got killed(you can feel it but cant be sure)

### PR DESCRIPTION
**Pull request by necropotame**
Could be faster if we just render 8 rectangles with a "death tile" texture, but we need 3D texture for that (see #675).

I can of course remove the blinking effect, but I think it's better to give a hint that those tiles are not really in the map

### CODE REVIEW

File: src/game/client/components/maplayers.cpp [hunk 1]
```diff
@@ -263,6 +263,21 @@ void CMapLayers::OnRender()
 						RenderTools()->RenderTilemap(pTiles, pTMap->m_Width, pTMap->m_Height, 32.0f, Color, TILERENDERFLAG_EXTEND|LAYERRENDERFLAG_OPAQUE,
 														EnvelopeEval, this, pTMap->m_ColorEnv, pTMap->m_ColorEnvOffset);
 						Graphics()->BlendNormal();
+						
+						//Draw kill tiles outside the entity clipping rectangle
```
Comment (heinrich5991) [inline src/game/client/components/maplayers.cpp]: [Style: Generally there's a space between `//` and the comment. `// Draw kill tiles …`.]

File: src/game/client/components/maplayers.cpp [hunk 2]
```diff
@@ -263,6 +263,21 @@ void CMapLayers::OnRender()
...
+						{
+							
+							//Slow blinking to hint that it's not a part of the map
+							double Seconds = time_get()/(double)time_freq();
+							vec4 ColorHint = vec4(1.0f, 1.0f, 1.0f, 0.3f + 0.7f*(1.0+sin(2.0f*pi*Seconds/3.f))/2.0f);
+							
+							RenderTools()->RenderTileRetangle(-201, -201, pTMap->m_Width+402, pTMap->m_Height+402,
+							                                  0, TILE_DEATH, //Display air inside, death outside
+							                                  32.0f, Color*ColorHint, TILERENDERFLAG_EXTEND|LAYERRENDERFLAG_TRANSPARENT,
+															  EnvelopeEval, this, pTMap->m_ColorEnv, pTMap->m_ColorEnvOffset);
```
Comment (heinrich5991) [inline src/game/client/components/maplayers.cpp]: [Style: Indent with tabs, align with spaces. This looks wrong with Github's 8 spaces per tab setting.]

File: src/game/client/render.h [hunk 3]
```diff
@@ -76,7 +76,8 @@ class CRenderTools
 	void RenderQuads(CQuad *pQuads, int NumQuads, int Flags, ENVELOPE_EVAL pfnEval, void *pUser);
 	void ForceRenderQuads(CQuad *pQuads, int NumQuads, int Flags, ENVELOPE_EVAL pfnEval, void *pUser, float Alpha = 1.0f);
 	void RenderTilemap(CTile *pTiles, int w, int h, float Scale, vec4 Color, int RenderFlags, ENVELOPE_EVAL pfnEval, void *pUser, int ColorEnv, int ColorEnvOffset);
-
+	void RenderTileRetangle(int X, int Y, int w, int h, unsigned char indexIn, unsigned char indexOut, float Scale, vec4 Color, int RenderFlags, ENVELOPE_EVAL pfnEval, void *pUser, int ColorEnv, int ColorEnvOffset);
```
Comment (heinrich5991) [inline src/game/client/render.h]: *`Rectangle`

File: src/game/client/render.h [hunk 4]
```diff
@@ -76,7 +76,8 @@ class CRenderTools
...
+
```
Comment (heinrich5991) [inline src/game/client/render.h]: [Style: Lines without content should not have whitespace.]

File: src/game/client/render_map.cpp [hunk 5]
```diff
@@ -176,6 +176,83 @@ void CRenderTools::ForceRenderQuads(CQuad *pQuads, int NumQuads, int RenderFlags
 	Graphics()->QuadsEnd();
 }
 
+void CRenderTools::RenderTileRetangle(int X, int Y, int w, int h, unsigned char indexIn, unsigned char indexOut, float Scale, vec4 Color, int RenderFlags,
```
Comment (heinrich5991) [inline src/game/client/render_map.cpp]: We shouldn't have variables whose names only differ by capitalization.

`indexIn` and `indexOut` do not adhere to the naming convention.

File: src/game/client/render_map.cpp [hunk 6]
```diff
@@ -176,6 +176,83 @@ void CRenderTools::ForceRenderQuads(CQuad *pQuads, int NumQuads, int RenderFlags
...
+									ENVELOPE_EVAL pfnEval, void *pUser, int ColorEnv, int ColorEnvOffset)
```
Comment (heinrich5991) [inline src/game/client/render_map.cpp]: Align with spaces, if you must. Cutting at 154 characters is a bit arbitrary. If you do want to break, consider breaking at 80 or 100 characters.

File: src/game/client/render_map.cpp [hunk 7]
```diff
@@ -176,6 +176,83 @@ void CRenderTools::ForceRenderQuads(CQuad *pQuads, int NumQuads, int RenderFlags
...
+
+	// adjust the texture shift according to mipmap level
+	float TexSize = 1024.0f;
+	float Frac = (1.25f/TexSize) * (1/FinalTilesetScale);
+	float Nudge = (0.5f/TexSize) * (1/FinalTilesetScale);
+
+	for(int y = StartY; y < EndY; y++)
+		for(int x = StartX; x < EndX; x++)
+		{
+			unsigned char Index = (x<X || x>=X+w || y<Y || y>Y+h) ? indexOut : indexIn;
```
Comment (heinrich5991) [inline src/game/client/render_map.cpp]: You should probably document what exactly this function includes in the rectangle and what it does not.

Currently this looks inconsistent, `x >= X + w` but `y > Y + h`?

File: src/game/client/render_map.cpp [hunk 8]
```diff
@@ -176,6 +176,83 @@ void CRenderTools::ForceRenderQuads(CQuad *pQuads, int NumQuads, int RenderFlags
...
+	int StartX = (int)(ScreenX0/Scale)-1;
+	int EndY = (int)(ScreenY1/Scale)+1;
+	int EndX = (int)(ScreenX1/Scale)+1;
+
+	// adjust the texture shift according to mipmap level
+	float TexSize = 1024.0f;
+	float Frac = (1.25f/TexSize) * (1/FinalTilesetScale);
+	float Nudge = (0.5f/TexSize) * (1/FinalTilesetScale);
+
+	for(int y = StartY; y < EndY; y++)
```
Comment (heinrich5991) [inline src/game/client/render_map.cpp]: [Style: While you're at it, you could probably add braces to this huge block.]

**Comment by necropotame** [comment]
Thanks for the review :)

**Comment by heinrich5991** [comment]
Looks good to me.

**Pull request merged**